test/gtest/ucs/test_datatype.cc

```diff
@@ -1356,3 +1359,215 @@ UCS_MT_TEST_F(test_datatype_ptr_map_safe, safe_put, num_threads)
         }
     }
 }
+
+
+class test_piecewise_func : public test_datatype {
+protected:
+    /* Represents piecewise function segment which consists of the linear
+     * function and the range on which this function is applicable.
+     * Both `start` and `end` is included to the range.
+     */
+    struct segment {
+        size_t            start;
+        size_t            end;
+        ucs_linear_func_t func;
+    };
+
+    static std::string get_segment_string(const segment &seg)
+    {
+        std::stringstream ss;
+        ss << "segment [" << seg.start << ", " << seg.end << "] : ";
+        ss << "func f(x) = " << seg.func.c << " + x*" << seg.func.m;
+
+        return ss.str();
+    }
+
+    static void print_piecewise_func(const ucs_piecewise_func_t &func)
+    {
+        size_t start = 0;
+        ucs_piecewise_segment_t *seg;
+
+        UCS_TEST_MESSAGE << "piecewise func:";
+        ucs_list_for_each(seg, &func.head, list) {
+            segment test_segment = {start, seg->end, seg->func};
+            UCS_TEST_MESSAGE << "\t" << get_segment_string(test_segment);
+            start = seg->end + 1;
+        }
+    }
+
+    static void compare_point_value(const ucs_piecewise_func_t &piecewise_func,
+                                    const ucs_linear_func_t &linear_func,
+                                    size_t point)
+    {
+        ASSERT_EQ(ucs_piecewise_func_apply(&piecewise_func, point),
+                  ucs_linear_func_apply(linear_func, point))
+                  << "point is " << point;
+    }
+
+    static size_t get_num_segments(ucs_piecewise_func_t &func)
+    {
+        return ucs_list_length(&func.head);
```

Comment:
IMO better add C API for this; avoid accessing func fields from the test

Reply:
Done.